python/lsst/ts/ess/common/sensor/campbell_csat3b.py

```diff
@@ -0,0 +1,170 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["compute_signature", "Csat3bSensor"]
+
+import logging
+import numpy as np
+
+from .base_sensor import BaseSensor
+from ..constants import SensorType
+from .sensor_registry import register_sensor
+
+_NUM_VALUES = 7
+
+
+def compute_signature(input_str: str, delimiter: str) -> int:
+    """Compute the signature of the sensor telemetry.
+
+    The computation is based on the C function that can be found on page
+    55 of csat3b.pdf in the doc directory.
+
+    Parameters
+    ----------
+    input_str : `str`
+        The string of telemetry for which to compute the signature.
+    delimiter: `str`
+        The delimiter used in input_str.
+
+    Returns
+    -------
+    `int`
+        The signature.
+
+    Notes
+    -----
+    By experiment it was found that the description of the input string on
+    page 55 of the document is incorrect. The description states that all
+    data from "x" to "c" are used but by implementing the C function in a C
+    source file and calling the function it was established that only "x"
+    up to "d" are used.
+    For compatibility sake, the input string is expected to contain the
+    value of "c" plus a leading delimiter and this Python implementation
+    strips that value and delimiter before computing the signature.
+    """
+    # Remove the value of "c" plus the leading delimiter from the input
+    # string otherwise an incorrect signature is computed.
+    last_index = input_str.rfind(delimiter)
+    input_str = input_str[:last_index]
+
+    # This is a Python version of the C function.
+    seed = 0xAAAA
+    msb = seed >> 8
+    # Ensure that the value of lsb is in the range of a C unsigned short.
+    lsb = seed % 256
+    b = 0
+    for char in input_str:
+        # Ensure that the value of b is in the range of a C unsigned short.
+        b = ((lsb << 1) + msb + ord(char)) % 256
+        if lsb & 0x80:
+            b = b + 1
+        msb = lsb
+        lsb = b
+    # Ensure that the return value is in the range of a C unsigned word.
+    return ((msb << 8) + lsb) % 65536
+
+
+class Csat3bSensor(BaseSensor):
+    """CSAT3B Sensor.
+
+    Perform protocol conversion for Campbell Scientific CSAT3B Anemometer
+    instruments. Serial data is output by the anemometer once per 100 ms with
+    the following format:
+
+        'X,Y,Z,T,D,C,S<CR>'
+
+    where:
+
+        X         x-axis wind speed (m/s). Decimal value.
+        Y         y-axis wind speed (m/s). Decimal value.
+        Z         z-axis wind speed (m/s). Decimal value.
+        T         Sonic temperature (degrees C). Decimal value.
+        D         Diagnostic word. Single digit decimal.
+        C         Record counter. One or two digit value (0-63).
+        S         Signature. Four character hexadecimal value.
```

Comment:
Including "0x" or not? Are a-e uppercase or lowercase? An example would be helpful -- either an example value of S here, or a full example return string that includes at least one a-f in the S value.

Reply:
I expanded the comment and referred to page 55 of the manual for examples.